Cannot select text within a label element that is linked to an input field

Ahmad-S792 opened the pull request:
#### 01527bde393728d7c127bcc2b777baa4397c75c0
<pre>
Cannot select text within a label element that is linked to an input field

<a href="https://bugs.webkit.org/show_bug.cgi?id=96734">https://bugs.webkit.org/show_bug.cgi?id=96734</a>

Reviewed by Aditya Keerthi.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=174296 &amp;
<a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=174818

Any click event on label goes to its control element. So, if
we tried to select the text inside label by dragging the mouse
or by double clicking on the text, it went to its control
element. This patch checks whether there is selection or not,
after the click is happened. If there is selection then do
nothing else continue.

If document is detached and click event is generated on
a label element, it causes a crash as the document doesn&apos;t
have frame. So adding additional check for LocalFrame.

* Source/WebCore/html/HTMLLabelElement.cpp:
(HTMLLabelElement::isEventTargetedAtInteractiveDescendants):
* LayoutTests/fast/forms/label/label-selection.html: Add Test Case
* LayoutTests/fast/forms/label/label-selection-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Skip test due to lack of MouseEvent support on iOS platform

Canonical link: <a href="https://commits.webkit.org/263420@main">https://commits.webkit.org/263420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87ec9d5333ffaf8d5efc8c8e39e75709f2b3e342

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3100 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3269 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4512 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3188 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2697 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3131 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3458 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2774 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4314 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/956 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2751 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2599 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2730 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2802 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4045 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2534 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2766 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2753 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1121 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2757 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3020 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/label/label-selection.html:
```diff
@@ -0,0 +1,29 @@
+    eventSender.mouseDown();
+    eventSender.mouseUp();
+    eventSender.mouseDown();
+    eventSender.mouseUp();
```
Let's use `UIHelper.doubleClickAt`.

Ahmad-S792 replied:
@pxlcoder - It seems it will take me some time to figure out this test case (not good with JS). Should we land updated test in separate patch?

pxlcoder replied:
You can add `<script src="../../../resources/ui-helper.js"></script>` on line 4.

And then replace this code with `UIHelper.doubleClickAt(element.offsetLeft, element.offsetTop)`.

Ahmad-S792 replied:
Seems like this is now good to go and thanks again for your help as always. Really appreciate it.. Now let's make next Safari release awesome..

pxlcoder commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -140,6 +143,15 @@ bool HTMLLabelElement::isEventTargetedAtInteractiveDescendants(Event& event) con
+        if (event.isMouseEvent() && !downcast<MouseEvent>(event).isSimulated()) {
+            if (auto* frame = document().frame()) {
+                if (frame->selection().selection().isRange())
+                return;
```
Nit: Incorrect indentation.

Ahmad-S792 replied:
Done.

pxlcoder commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -140,6 +143,15 @@ bool HTMLLabelElement::isEventTargetedAtInteractiveDescendants(Event& event) con
+        if (event.isMouseEvent() && !downcast<MouseEvent>(event).isSimulated()) {
+            if (auto* frame = document().frame()) {
+                if (frame->selection().selection().isRange())
+                return;
```
Might be more readable as:

```
auto* frame = document().frame();
if (frame && frame->selection().selection().isRange())
    return;
```

To reduce the nesting.

Ahmad-S792 replied:
Done.

pxlcoder commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -140,6 +143,14 @@ bool HTMLLabelElement::isEventTargetedAtInteractiveDescendants(Event& event) con
+        // Note: a click event may be not a mouse event if created by document.createEvent().
+        if (event.isMouseEvent() && !downcast<MouseEvent>(event).isSimulated()) {
+            auto* frame = document().frame();
+            if (frame && frame->selection().selection().isRange())
```
Can you test the case where the label is already selected, and then you click on the selected text? Would like to make sure the control element is still focused in that case.

Ahmad-S792 replied:
I found the desired test in Chrome / Blink here - [Link](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/fast/forms/label/label-selection-by-textSelection-and-click.html) but it does not work in WebKit and similarly, I created this pre-selected label text fiddle ([link](https://jsfiddle.net/dao3m021/1/)), when double clicking non-patched system, I am unable to get text selection with double click while after patched local, I am able to get selection. (Single click still moves to Input focus similar to Chrome Canary).

Ahmad-S792 replied:
@pxlcoder - Anything on above from your side?

pxlcoder replied:
> Single click still moves to Input focus similar to Chrome Canary

Sounds good!

Ahmad-S792 committed:
Cannot select text within a label element that is linked to an input field

https://bugs.webkit.org/show_bug.cgi?id=96734

Reviewed by Aditya Keerthi.

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=174296 &
https://src.chromium.org/viewvc/blink?view=revision&revision=174818

Any click event on label goes to its control element. So, if
we tried to select the text inside label by dragging the mouse
or by double clicking on the text, it went to its control
element. This patch checks whether there is selection or not,
after the click is happened. If there is selection then do
nothing else continue.

If document is detached and click event is generated on
a label element, it causes a crash as the document doesn't
have frame. So adding additional check for LocalFrame.

* Source/WebCore/html/HTMLLabelElement.cpp:
(HTMLLabelElement::isEventTargetedAtInteractiveDescendants):
* LayoutTests/fast/forms/label/label-selection.html: Add Test Case
* LayoutTests/fast/forms/label/label-selection-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Skip test due to lack of MouseEvent support on iOS platform

Canonical link: https://commits.webkit.org/263420@main
